[css-view-transitions] Prevent view transitions when a swipe animation is already running.

mattwoodrow opened the pull request:
#### 33ad37aabfeee530c2a37e445fc32f5b9c524a8c
<pre>
[css-view-transitions] Prevent view transitions when a swipe animation is already running.
<a href="https://bugs.webkit.org/show_bug.cgi?id=285886">https://bugs.webkit.org/show_bug.cgi?id=285886</a>
&lt;<a href="https://rdar.apple.com/142844150">rdar://142844150</a>&gt;

Reviewed by Tim Nguyen.

Prevent cross-document view-transitions if a swipe animation is already running, as recommended by the spec.

Also skip the view transition if the navigation was initiated from the UI (and isn&apos;t a history traversal).

Added new test that is identical to basic-cached-back-swipe.html, with the addition of a view transitions check.

* LayoutTests/swipe/swipe-disables-view-transition-expected.txt: Added.
* LayoutTests/swipe/swipe-disables-view-transition.html: Added.
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::navigationCanTriggerCrossDocumentViewTransition):

Canonical link: <a href="https://commits.webkit.org/288956@main">https://commits.webkit.org/288956@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d1ffde48244b5bac0df42ad9883732f55ada8adf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84746 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4471 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39134 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89885 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35798 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86831 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4560 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12446 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65949 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23774 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87791 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3460 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77014 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46223 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3339 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31233 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34872 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74189 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32039 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91261 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12085 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8822 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74424 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12312 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72825 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73550 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18230 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17926 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16372 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/3533 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12037 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17477 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11871 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15365 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13617 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (approved):
LGTM but would be nice to have a test

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1824,6 +1824,12 @@ void FrameLoader::loadWithDocumentLoader(DocumentLoader* loader, FrameLoadType t
+        m_loadStartedDuringSwipeAnimation = page->isInSwipeAnimation();
+    else
+        m_loadStartedDuringSwipeAnimation = false;
+
```
nit: extra space

achristensen07 commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -2515,7 +2515,8 @@ ShouldOpenExternalURLsPolicy DocumentLoader::shouldOpenExternalURLsPolicyToPropa
 bool DocumentLoader::navigationCanTriggerCrossDocumentViewTransition(Document& oldDocument, bool fromBackForwardCache)
 {
-    // FIXME: Consider adding implementation-defined navigation experience step.
+    if (frameLoader()->loadStartedDuringSwipeAnimation())
```
May as well null check the frame loader.

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1824,6 +1824,12 @@ void FrameLoader::loadWithDocumentLoader(DocumentLoader* loader, FrameLoadType t
+    if (RefPtr page = frame->page())
+        m_loadStartedDuringSwipeAnimation = page->isInSwipeAnimation();
+    else
+        m_loadStartedDuringSwipeAnimation = false;
```
Does this state need to persist after the animations and transitions are done?  I think this state would be better on the DocumentLoader because the main Frame can last many navigations but the DocumentLoader seems to have a more similar scope to the lifetime of this bool.

mattwoodrow replied:
It does not, so have moved the bool to the DocumentLoader!

mattwoodrow committed:
[css-view-transitions] Prevent view transitions when a swipe animation is already running.
https://bugs.webkit.org/show_bug.cgi?id=285886
<rdar://142844150>

Reviewed by Tim Nguyen.

Prevent cross-document view-transitions if a swipe animation is already running, as recommended by the spec.

Also skip the view transition if the navigation was initiated from the UI (and isn't a history traversal).

Added new test that is identical to basic-cached-back-swipe.html, with the addition of a view transitions check.

* LayoutTests/swipe/swipe-disables-view-transition-expected.txt: Added.
* LayoutTests/swipe/swipe-disables-view-transition.html: Added.
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::navigationCanTriggerCrossDocumentViewTransition):

Canonical link: https://commits.webkit.org/288956@main
